seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/logical/LogicalDagOptimizer.java

```diff
@@ -0,0 +1,180 @@
+
+    private final List<LogicalVertex> logicalJobVertices;
+
+    private final Set<LogicalEdge> oldEdges;
+
+    /**
+     * key: old vertex id, value: new vertex id;
+     */
+    private final Map<Long, Long> oldToNewIdMap = new HashMap<>();
+
+    private final LogicalDag dag;
+
+    public LogicalDagOptimizer(final Map<Long, LogicalVertex> logicalVertexMap,
+                               final Set<LogicalEdge> edges,
+                               JobConfig jobConfig) {
+        this.newLogicalVertexMap = new HashMap<>(logicalVertexMap.size());
+        this.logicalJobVertices = new ArrayList<>(logicalVertexMap.values());
+        this.oldEdges = edges;
+        this.dag = new LogicalDag(jobConfig, idGenerator);
+    }
+
+    public LogicalDag expandAndOptimize() {
+        getSourceVertices().forEach(sourceVertex -> {
+            List<LogicalVertex> vertices = new ArrayList<>();
+            vertices.add(sourceVertex);
+            for (int index = 0; index < vertices.size(); index++) {
+                LogicalVertex vertex = vertices.get(index);
+                createChainedVertex(vertex);
+                vertex.getOutputs().forEach(edge -> vertices.add(edge.getTargetVertex()));
+            }
+        });
+        connectVertices();
+        dag.getLogicalVertexMap().putAll(newLogicalVertexMap);
+        return dag;
+    }
+
+    private void connectVertices() {
+        oldEdges.forEach(oldEdge -> {
+            Long oldInputVertexId = oldEdge.getInputVertex().getVertexId();
+            Long oldTargetVertexId = oldEdge.getTargetVertex().getVertexId();
+            LogicalVertex inputVertex = newLogicalVertexMap.get(oldToNewIdMap.get(oldInputVertexId));
+            LogicalVertex targetVertex = newLogicalVertexMap.get(oldToNewIdMap.get(oldTargetVertexId));
+            LogicalEdge edge = new LogicalEdge(inputVertex, targetVertex);
+            inputVertex.addOutput(targetVertex);
+            targetVertex.addInput(inputVertex);
+            dag.addEdge(edge);
+        });
+    }
+
+    private List<LogicalVertex> getSourceVertices() {
+        List<LogicalVertex> vertices = new ArrayList<>();
+        for (LogicalVertex vertex : logicalJobVertices) {
+            if (vertex.getInputs().size() == 0) {
+                vertices.add(vertex);
+            }
+        }
+        return vertices;
+    }
+
+    private void createChainedVertex(LogicalVertex logicalVertex) {
+        if (oldToNewIdMap.containsKey(logicalVertex.getVertexId())) {
+            return;
+        }
+        final ArrayList<LogicalVertex> chainedVertices = new ArrayList<>();
+        chainedVertices(logicalVertex, chainedVertices);
+        final long newId = idGenerator.getNextId();
+        Action newAction;
+        if (chainedVertices.size() < 1) {
+            Action action = logicalVertex.getAction();
+            if (action instanceof PartitionTransformAction) {
+                newAction = new PartitionTransformAction(newId,
```

Comment:
newAction is same with action, why not just set `newAction = action`?

Reply:
Because it need to change the action ID.